README
Term-ReadLine::Tiny
===================


INFO

Read a line.



INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install



SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Term::ReadLine::Tiny



LICENSE AND COPYRIGHT

Copyright (C) 2012-2014 Matthäus Kiem

This library is free software; you
can redistribute it and/or modify it under the same terms
as Perl 5.10.0. For details, see the full text of the
licenses in the file LICENSE.


Changes
Revision history for Term::ReadLine::Tiny


0.011   2014-08-27
        - The module is now called "Term::ReadLine::Simple". "Term::ReadLine::Tiny" will be removed".

0.010_05   2014-08-24
        - Perl minimum version from "5.10.0" to "5.08.0".
        - Code refactoring.
        - Update documentation.

0.010_04   2014-08-17
        - Removed non-ascii. (srezic - rt 98094)

0.010_03   2014-08-16
        - Update test.

0.010_02   2014-08-16
        - New tests.

0.010_01   2014-08-15
        - Added "Ctrl+K" and changed "Ctrl+U".
        - New tests.
        - Update documentation.

0.010   2014-08-14
        - Fixed bug in "config".
        - Update keycodes.
        - Code refactoring.
        - Perl minimum version is now "5.10.0".
        - New test.

0.009   2014-07-31
        - Code refactoring: removed "\e[6n"-dependency.

0.008   2014-07-28
        - Bugfix.
        - Update Makefile.PL.
        - Update documentation.

0.007   2014-07-28
        - Code refactoring.

0.006   2014-07-27
        - MSWin32: replaced ANSI escape sequences.

0.005   2014-07-27
        - Bugfixes.
        - MSWin32: removed dependency on "Win32::Console::ANSI".
        - Update documentation.

0.004   2014-07-26
        - Bugfixes.
        - Removed the key-support-limitation for MSWin32.
        - Removed option "asterix".
        - Code refactoring.
        - Update documentation.

0.003   2014-07-24
        - New keys.
        - "readline" returns always decoded strings if not set "compat" or "$ENV{READLINE_TINY_COMPAT}".
        - Update documentation.

0.002   2014-07-16
        - Changed the handling of the decoding.
        - Added option "compat" and env variable "READLINE_TINY_COMPAT".
        - Update documentation.

0.001   2014-07-14
        - First release on CPAN.


t/readline.pl
#!/usr/bin/env perl
use warnings;
use strict;
use 5.008000;

use Term::ReadLine::Tiny;

use FindBin qw( $RealBin );
use lib $RealBin;
use Data_Test_Readline;

my $tiny = Term::ReadLine::Tiny->new();
my $a_ref = Data_Test_Readline::return_test_data();

for my $ref ( @$a_ref ) {
    my $args  = $ref->{arguments};

    my $line = $tiny->readline( @$args );
    print "<$line>\n";
}


t/compat_readline.pl
#!/usr/bin/env perl
use warnings;
use strict;
use 5.010000;
binmode STDOUT, ':utf8';
binmode STDIN, ':utf8';

use Term::ReadLine::Tiny;

use FindBin qw( $RealBin );
use lib $RealBin;
use Data_Test_Readline;

my $a_ref  = Data_Test_Readline::return_test_data();

my $tiny = Term::ReadLine::Tiny->new();
$tiny->config( { compat => 1 } );

for my $ref ( @$a_ref ) {
    my $args  = $ref->{arguments};

    my $line = $tiny->readline( @$args );
    say "<$line>";
}


t/Data_Test_Readline.pm
package # hide
Data_Test_Readline;

use 5.010000;
use warnings;
use strict;


sub key_seq {
    return {
        CONTROL_M => "\x{0d}",
        ENTER     => "\x{0d}",

        RIGHT     => "\e[C",
        RIGHT_O   => "\eOC",
        CONTROL_F => "\x{06}",

        LEFT      => "\e[D",
        LEFT_O    => "\eOD",
        CONTROL_B => "\x{02}",

        CONTROL_A => "\x{01}",
        HOME      => "\e[H",

        CONTROL_E => "\x{05}",
        END       => "\e[F",

        CONTROL_H => "\x{08}",
        BTAB      => "\x{08}",
        BSPACE    => "\x{7f}",
        BTAB_Z    => "\e[Z",
        BTAB_OZ   => "\eOZ",

        CONTROL_D => "\x{04}",
        DELETE    => "\e[3~",

        CONTROL_K => "\x{0b}",
        CONTROL_U => "\x{15}",
    };
}

sub return_test_data {
    return [
        {
            used_keys => [ "-", 'ENTER' ],
            expected  => "<->",
            arguments => [ 'Prompt: ' ],
        },
        {
            used_keys => [ 'ENTER' ],
            expected  => "<default>",
            arguments => [ 'Prompt: ', "default" ],
        },
        {
            used_keys => [ ( 'LEFT' ) x 4, 'CONTROL_K', " ", 'ENTER' ],
            expected  => "<def >",
            arguments => [ 'Prompt: ', 'default' ],
        },
        {
            used_keys => [ ( 'LEFT' ) x 5, "-", 'ENTER' ],
            expected  => "<de-fault>",
            arguments => [ 'Prompt: ', "default" ],
        },
        {
            used_keys => [ ( 'BSPACE' ) x 7, "hello", " ", "world", 'ENTER' ],
            expected  => "<hello world>",
            arguments => [ 'Prompt: ', { default => "default" } ],
        },
        {
            used_keys => [ "house", 'ENTER' ],
            expected  => "<The house>",
            arguments => [ 'Prompt: ', { default => "The ", no_echo => 1 } ],
        },
        {
            used_keys => [ "0123456789", ( 'LEFT' ) x 5, 'CONTROL_U', 'ENTER' ],
            expected  => "<56789>",
            arguments => [ ': ' ],
        },
        {
             used_keys => [ 'CONTROL_U', "a", 'ENTER'],
             expected  => "<a>",
             arguments => [ 'Prompt: ', { default => "many words " x 1000, no_echo => 0 } ],
        },
        {
            used_keys => [ "abcde ", "XY " x 10, ( 'LEFT' ) x 36, ( 'RIGHT' ) x 8, 'CONTROL_K', 'ENTER' ],
            expected  => "<DEFAULT abcde XY>",
            arguments => [ "Prompt: ", { default => "DEFAULT " } ],
        },
        {
            used_keys => [ "The black cat climbed the green tree", 'HOME', ( 'RIGHT' ) x 4, ( 'DELETE' ) x 6, 'END', ( 'LEFT' ) x 5, ( 'BSPACE' ) x 6, 'ENTER' ],
            expected  => "<The cat climbed the tree>",
            arguments => [ 'Prompt: ' ],
        },
        {
            used_keys => [ qw( HOME RIGHT RIGHT_O DELETE CONTROL_D END LEFT LEFT_O BSPACE CONTROL_H CONTROL_A DELETE CONTROL_E BSPACE _ABC ENTER ) ],
            expected  => "<hblack cat climbed the green e_ABC>",
            arguments => [ 'Prompt: ', { default => "The black cat climbed the green tree" } ],
        },
    ];
}




1;

__END__


Makefile.PL
use 5.008000;
use warnings;
use strict;
use ExtUtils::MakeMaker;


my %prereqs_os_specific;
my %test_prereqs_os_specific;
if ( $^O eq 'MSWin32' ) {
    %prereqs_os_specific = (
        'Win32::Console' => 0,
    );
}
else {
    %prereqs_os_specific = (
        'Term::ReadKey'  => 0,
    );
    %test_prereqs_os_specific = (
        'Expect'         => 0,
    );
}


WriteMakefile(
    PL_FILES         => {},
    MIN_PERL_VERSION => '5.008000',
    LICENSE          => 'perl',
    META_MERGE => {
        'meta-spec' => { version => 2 },
        resources   => {
            repository => {
              type => 'git',
              web  => 'https://github.com/kuerbis/Term-ReadLine::Tiny',
              url  => 'https://github.com/kuerbis/Term-ReadLine::Tiny.git',
            },
        },
    },
    AUTHOR           => 'Matthaeus Kiem <cuer2s@gmail.com>',
    NAME             => 'Term::ReadLine::Tiny',
    ABSTRACT_FROM    => 'lib/Term/ReadLine/Tiny.pm',
    VERSION_FROM     => 'lib/Term/ReadLine/Tiny.pm',
    EXE_FILES        => [],
    BUILD_REQUIRES   => {
        'Test::More'        => 0,
        'Test::Fatal'       => 0,
        %test_prereqs_os_specific,
    },
    PREREQ_PM => {
        'strict'            => 0,
        'warnings'          => 0,
        'Carp'              => 0,
        'Encode'            => 0,
        'Encode::Locale'    => 0,
        'Exporter'          => 0,
        'Unicode::GCString' => 0,
        %prereqs_os_specific,
    },
    dist  => {
        COMPRESS => 'gzip',
        SUFFIX   => '.gz',
        TARFLAGS => '--format=gnu -cvf',
    },
    clean => {
        FILES => 'Term-ReadLine-Tiny-*',
    },
);


lib/Term/ReadLine/Tiny/Linux.pm
package # hide from PAUSE
Term::ReadLine::Tiny::Linux;

use warnings;
use strict;
use 5.008000;

our $VERSION = '0.011';

use Term::ReadKey  qw( GetTerminalSize ReadKey ReadMode );

use Term::ReadLine::Tiny::Constants qw( :linux );


sub new {
    return bless {}, $_[0];
}


sub __set_mode {
    #my ( $self ) = @_;
    ReadMode( 'cbreak' );
};


sub __reset_mode {
    #my ( $self ) = @_;
    ReadMode( 'restore' );
}


sub __term_buff_width {
    #my ( $self ) = @_;
    my ( $term_width ) = GetTerminalSize();
    return $term_width;
}


sub __get_key {
    #my ( $self ) = @_;
    my $c1 = ReadKey( 0 );
    return if ! defined $c1;
    if ( $c1 eq "\e" ) {
        my $c2 = ReadKey( 0.10 );
        if ( ! defined $c2 ) {
            return  NEXT_get_key; # KEY_ESC
        }
        elsif ( $c2 eq 'O' ) {
            my $c3 = ReadKey( 0 );
               if ( $c3 eq 'C' ) { return VK_RIGHT; }
            elsif ( $c3 eq 'D' ) { return VK_LEFT; }
            elsif ( $c3 eq 'F' ) { return VK_END; }
            elsif ( $c3 eq 'H' ) { return VK_HOME; }
            elsif ( $c3 eq 'Z' ) { return KEY_BTAB; }
            else {
                return NEXT_get_key;
            }
        }
        elsif ( $c2 eq '[' ) {
            my $c3 = ReadKey( 0 );
               if ( $c3 eq 'C' ) { return VK_RIGHT; }
            elsif ( $c3 eq 'D' ) { return VK_LEFT; }
            elsif ( $c3 eq 'F' ) { return VK_END; }
            elsif ( $c3 eq 'H' ) { return VK_HOME; }
            elsif ( $c3 eq 'Z' ) { return KEY_BTAB; }
            elsif ( $c3 =~ /^[0-9]$/ ) {
                my $c4 = ReadKey( 0 );
                if ( $c4 eq '~' ) {
                    if ( $c3 eq '3' ) { return VK_DELETE; }
                    else {
                        return NEXT_get_key;
                    }
                }
                else {
                    return NEXT_get_key;
                }
            }
            else {
                return NEXT_get_key;
            }
        }
        else {
            return NEXT_get_key;
        }
    }
    else {
        return ord $c1;
    }
};


sub __up    { print "\e[${_[1]}A"; }


sub __left  { print "\e[${_[1]}D"; }


sub __right { print "\e[${_[1]}C"; }


sub __clear_output { print "\e[0J"; }


sub __save_cursor_position    { print "\e[s"; }


sub __restore_cursor_position { print "\e[u"; }



1;

__END__


lib/Term/ReadLine/Tiny/Win32.pm
package # hide from PAUSE
Term::ReadLine::Tiny::Win32;

use warnings;
use strict;
use 5.008000;

our $VERSION = '0.011';

use Encode qw( decode );

use Encode::Locale qw();
use Win32::Console qw( STD_INPUT_HANDLE ENABLE_PROCESSED_INPUT STD_OUTPUT_HANDLE
                       RIGHT_ALT_PRESSED LEFT_ALT_PRESSED RIGHT_CTRL_PRESSED LEFT_CTRL_PRESSED SHIFT_PRESSED );

use Term::ReadLine::Tiny::Constants qw( :win32 );


sub new {
    return bless {}, $_[0];
}


sub __set_mode {
    my ( $self ) = @_;
    $self->{input} = Win32::Console->new( STD_INPUT_HANDLE );
    $self->{old_in_mode} = $self->{input}->Mode();
    $self->{input}->Mode( ENABLE_PROCESSED_INPUT );
    $self->{output} = Win32::Console->new( STD_OUTPUT_HANDLE );
    $self->{def_attr}  = $self->{output}->Attr();
    $self->{bg_color}  = $self->{def_attr} & 0x70;
    $self->{fill_attr} = $self->{bg_color} | $self->{bg_color};
}


sub __reset_mode {
    my ( $self ) = @_;
    if ( defined $self->{input} ) {
        if ( defined $self->{old_in_mode} ) {
            $self->{input}->Mode( $self->{old_in_mode} );
            delete $self->{old_in_mode};
        }
        $self->{input}->Flush;
        # workaround Bug #33513:
        delete $self->{input}{handle};
        delete $self->{output}{handle};
    }
}


sub SHIFTED_MASK () {
      RIGHT_ALT_PRESSED
    | LEFT_ALT_PRESSED
    | RIGHT_CTRL_PRESSED
    | LEFT_CTRL_PRESSED
    | SHIFT_PRESSED
}

sub __get_key {
    my ( $self ) = @_;
    my @event = $self->{input}->Input;
    my $event_type = shift @event;
    return NEXT_get_key if ! defined $event_type;
    if ( $event_type == 1 ) {
        my ( $key_down, $repeat_count, $v_key_code, $v_scan_code, $char, $ctrl_key_state ) = @event;
        return NEXT_get_key if ! $key_down;
        if ( $char ) {
            return ord decode( 'console_in', chr( $char & 0xff ) );
        }
        else{
            if ( $ctrl_key_state & SHIFTED_MASK ) {
                return NEXT_get_key;
            }
            elsif ( $v_key_code == VK_CODE_END )    { return VK_END }
            elsif ( $v_key_code == VK_CODE_HOME )   { return VK_HOME }
            elsif ( $v_key_code == VK_CODE_LEFT )   { return VK_LEFT }
            elsif ( $v_key_code == VK_CODE_UP )     { return VK_UP }
            elsif ( $v_key_code == VK_CODE_RIGHT )  { return VK_RIGHT }
            elsif ( $v_key_code == VK_CODE_DELETE ) { return VK_DELETE }
            else {
                return NEXT_get_key;
            }
        }
    }
    else {
        return NEXT_get_key;
    }
}


sub __term_buff_width {
    my ( $self ) = @_;
    my ( $term_width ) = $self->{output}->MaxWindow();
    return $term_width;
}


sub __get_cursor_position {
    my ( $self ) = @_;
    my ( $col, $row ) = $self->{output}->Cursor();
    return $col + 1, $row + 1;
}


sub __set_cursor_position {
    my ( $self, $col, $row ) = @_;
    $self->{output}->Cursor( $col - 1, $row - 1 );
}


sub __up {
    my ( $self, $rows_up ) = @_;
    return if ! $rows_up; #
    my ( $col, $row ) = $self->__get_cursor_position;
    my $new_row = $row - $rows_up;
    $new_row = 1 if $new_row < 1;
    $self->__set_cursor_position( $col, $new_row  );
}


sub __clear_output{
    my ( $self, $chars ) = @_;
    my ( $col, $row ) = $self->__get_cursor_position();
    $self->{output}->FillAttr(
            $self->{fill_attr},
            $chars,
            $col - 2, $row - 1 );
}


sub __save_cursor_position {
    my ( $self ) = @_;
    $self->{saved_cursor_position} = [ $self->{output}->Cursor() ];
}

sub __restore_cursor_position {
    my ( $self ) = @_;
    $self->{output}->Cursor( @{$self->{saved_cursor_position}} );
}




1;

__END__


lib/Term/ReadLine/Tiny/Constants.pm
package # hide from PAUSE
Term::ReadLine::Tiny::Constants;

use warnings;
use strict;
use 5.008000;

our $VERSION = '0.011';

use Exporter qw( import );

our @EXPORT_OK = qw(
        NEXT_get_key
        CONTROL_A CONTROL_B CONTROL_D CONTROL_E CONTROL_F CONTROL_H CONTROL_K CONTROL_U KEY_BTAB KEY_TAB
        KEY_ENTER KEY_ESC KEY_BSPACE
        VK_CODE_END VK_CODE_HOME VK_CODE_LEFT VK_CODE_UP VK_CODE_RIGHT VK_CODE_DELETE
        VK_END VK_HOME VK_LEFT VK_UP VK_RIGHT VK_DELETE
);

our %EXPORT_TAGS = (
    rl => [ qw(
        NEXT_get_key
        CONTROL_A CONTROL_B CONTROL_D CONTROL_E CONTROL_F CONTROL_H CONTROL_K CONTROL_U KEY_BTAB KEY_TAB
        KEY_ENTER KEY_ESC KEY_BSPACE
        VK_END VK_HOME VK_LEFT VK_UP VK_RIGHT VK_DELETE
    ) ],
    linux  => [ qw(
        NEXT_get_key
        KEY_BTAB KEY_ESC
        VK_END VK_HOME VK_LEFT VK_UP VK_RIGHT VK_DELETE
    ) ],
    win32  => [ qw(
        NEXT_get_key
        VK_CODE_END VK_CODE_HOME VK_CODE_LEFT VK_CODE_UP VK_CODE_RIGHT VK_CODE_DELETE
        VK_END VK_HOME VK_LEFT VK_UP VK_RIGHT VK_DELETE
    ) ]
);



#sub UP                     () { "\e[A" }
#sub RIGHT                  () { "\e[C" }
#sub LEFT                   () { "\e[D" }
#sub LF                     () { "\n" } #
#sub CR                     () { "\r" } #

#sub BEEP                   () { "\a" }
#sub CLEAR_SCREEN           () { "\e[2J\e[1;1H" }
#sub CLEAR_TO_END_OF_SCREEN () { "\e[0J" }

#sub SAVE_CURSOR_POSITION    () { "\e[s" }
#sub RESTORE_CURSOR_POSITION () { "\e[u" }

#sub GET_CURSOR_POSITION     () { "\e[6n" }


sub NEXT_get_key  () { -1 }


sub CONTROL_A  () { 0x01 }
sub CONTROL_B  () { 0x02 }
sub CONTROL_D  () { 0x04 }
sub CONTROL_E  () { 0x05 }
sub CONTROL_F  () { 0x06 }
sub CONTROL_H  () { 0x08 }
sub KEY_BTAB   () { 0x08 }
sub KEY_TAB    () { 0x09 }
sub CONTROL_K  () { 0x0b }
sub KEY_ENTER  () { 0x0d }
sub CONTROL_U  () { 0x15 }
sub KEY_ESC    () { 0x1b }
sub KEY_BSPACE () { 0x7f }


sub VK_END    () { 335 }
sub VK_HOME   () { 336 }
sub VK_LEFT   () { 337 }
sub VK_UP     () { 338 }
sub VK_RIGHT  () { 339 }
sub VK_DELETE () { 346 }


sub VK_CODE_END    () { 35 }
sub VK_CODE_HOME   () { 36 }
sub VK_CODE_LEFT   () { 37 }
sub VK_CODE_UP     () { 38 }
sub VK_CODE_RIGHT  () { 39 }
sub VK_CODE_DELETE () { 46 }

1;

__END__


lib/Term/ReadLine/Tiny.pm
package Term::ReadLine::Tiny;

use warnings;
use strict;
use 5.008000;

our $VERSION = '0.011';

use Carp   qw( croak carp );
use Encode qw( encode decode );

use Encode::Locale    qw();
use Unicode::GCString qw();

use Term::ReadLine::Tiny::Constants qw( :rl );

my $Plugin_Package;

BEGIN {
    if ( $^O eq 'MSWin32' ) {
        require Term::ReadLine::Tiny::Win32;
        $Plugin_Package = 'Term::ReadLine::Tiny::Win32';
    }
    else {
        require Term::ReadLine::Tiny::Linux;
        $Plugin_Package = 'Term::ReadLine::Tiny::Linux';
    }
}

sub ReadLine { 'Term::ReadLine::Tiny' }
sub IN {}
sub OUT {}
sub MinLine {}
sub Attribs { {} }
sub Features { { no_features => 1 } }
sub addhistory {}
sub ornaments {}


sub new {
    my $class = shift;
    my ( $name ) = @_;
    my $self = bless {
        name => $name,
    }, $class;
    $self->__set_defaults();
    $self->{plugin} = $Plugin_Package->new();
    return $self;
}


sub DESTROY {
    my ( $self ) = @_;
    $self->__reset_term();
}


sub __set_defaults {
    my ( $self ) = @_;
    #$self->{compat}          = undef if ! defined $self->{compat};
    #$self->{reinit_encoding} = undef if ! defined $self->{reinit_encoding};
    $self->{default} = '' if ! defined $self->{default};
    $self->{no_echo} = 0  if ! defined $self->{no_echo};
}


sub __validate_options {
    my ( $self, $opt ) = @_;
    return if ! defined $opt;
    my $valid = {
        no_echo         => '[ 0 1 2 ]',
        compat          => '[ 0 1 ]',
        reinit_encoding => '',
        default         => '',
    };
    my $sub =  ( caller( 1 ) )[3];
    $sub =~ s/^.+::([^:]+)\z/$1/;
    for my $key ( keys %$opt ) {
        if ( ! exists $valid->{$key} ) {
            croak $sub . ": '$key' is not a valid option name";
        }
        next if ! defined $opt->{$key};
        if ( ref $opt->{$key} ) {
            croak $sub . ": option '$key' : a reference is not a valid value.";
        }
        next if $valid->{$key} eq '';
        if ( $opt->{$key} !~ m/^$valid->{$key}\z/x ) {
            croak $sub . ": option '$key' : '$opt->{$key}' is not a valid value.";
        }
    }
}


sub __init_term {
    my ( $self ) = @_;
    $self->{plugin}->__set_mode();
    if ( $self->{reinit_encoding} ) {
        Encode::Locale::reinit( $self->{reinit_encoding} );
    }

}


sub __reset_term {
    my ( $self ) = @_;
    if ( defined $self->{plugin} ) {
        $self->{plugin}->__reset_mode();
    }
}


sub config {
    my ( $self, $opt ) = @_;
    if ( defined $opt ) {
        croak "config: the (optional) argument must be a HASH reference" if ref $opt ne 'HASH';
        $self->__validate_options( $opt );
        for my $option ( keys %$opt ) {
            $self->{$option} = $opt->{$option};
        }
    }
}


sub readline {
    my ( $self, $prompt, $opt ) = @_;
    if ( defined $prompt ) {
        croak "readline: a reference is not a valid prompt." if ref $prompt;
    }
    else {
        $prompt = '';
    }
    if ( defined $opt ) {
        if ( ! ref $opt ) {
            $opt = { default => $opt };
        }
        elsif ( ref $opt ne 'HASH' ) {
            croak "readline: the (optional) second argument must be a string or a HASH reference";
        }
        else {
            $self->__validate_options( $opt );
        }
    }
    else {
        $opt = {};
    }
    $opt->{default} = $self->{default} if ! defined $opt->{default};
    $opt->{no_echo} = $self->{no_echo} if ! defined $opt->{no_echo};
    my $gcs_prompt    = Unicode::GCString->new( $prompt );
    my $length_prompt = $gcs_prompt->length();
    my $str     = Unicode::GCString->new( $prompt . $opt->{default} );
    my $pos_str = $str->length();
    $self->{prev_str_cols} = $str->columns();

    local $| = 1;
    $self->__init_term();
    $self->{plugin}->__save_cursor_position();
    $self->__print_readline( $opt, $prompt, $str, $pos_str );

    while ( 1 ) {
        my $key = $self->{plugin}->__get_key();
        if ( ! defined $key ) {
            $self->__reset_term();
            carp "EOT: $!";
            return;
        }
        next if $key == NEXT_get_key;
        next if $key == KEY_TAB;
        if ( $key == KEY_BSPACE || $key == CONTROL_H ) {
            if ( $pos_str - $length_prompt ) {
                $pos_str--;
                $str->substr( $pos_str, 1, '' );
            }
        }
        elsif ( $key == CONTROL_U ) {
            $str->substr( $length_prompt, $pos_str - $length_prompt, '' );
            $pos_str = $length_prompt;
        }
        elsif ( $key == CONTROL_K ) {
            $str->substr( $pos_str, $str->length() - $pos_str, '' );
        }
        elsif ( $key == VK_DELETE || $key == CONTROL_D ) {
            if ( $str->length() - $length_prompt ) {
                if ( $pos_str < $str->length() ) {
                    $str->substr( $pos_str, 1, '' );
                }
            }
            else {
                print "\n";
                $self->__reset_term();
                return;
            }
        }
        elsif ( $key == VK_RIGHT || $key == CONTROL_F ) {
            $pos_str++ if $pos_str < $str->length();
        }
        elsif ( $key == VK_LEFT  || $key == CONTROL_B ) {
            if ( $pos_str > $length_prompt ) {
                $self->{manually_right} = 1 if $pos_str == $str->length();
                $pos_str--;
            }
        }
        elsif ( $key == VK_END   || $key == CONTROL_E ) {
            $pos_str = $str->length();
        }
        elsif ( $key == VK_HOME  || $key == CONTROL_A ) {
            $pos_str = $length_prompt;
        }
        else {
            $key = chr $key;
            utf8::upgrade $key;
            if ( $key eq "\n" || $key eq "\r" ) { ###
                $str->substr( 0, $length_prompt, '' );
                print "\n";
                $self->__reset_term();
                if ( $self->{compat} || ! defined $self->{compat} && $ENV{READLINE_TINY_COMPAT} ) {
                    return encode( 'console_in', $str->as_string );
                }
                return $str->as_string;
            }
            else {
                $str->substr( $pos_str, 0, $key );
                $pos_str++;
            }
        }
        $self->__print_readline( $opt, $prompt, $str, $pos_str );
    }
}


sub __print_readline {
    my ( $self, $opt, $prompt, $str, $pos_str ) = @_;
    my $tmp_pos = $str->pos();
    $str->pos( 0 );
    my $str_col = 0;
    my $str_row = 0;
    my @gc_in_row = ();
    my $term_width = $self->{plugin}->__term_buff_width();
    while ( defined( my $gc = $str->next ) ) {
        if ( $term_width < ( $str_col += $gc->columns ) ) {
            $str_col = $gc->columns();
            $str_row++;
        }
        $gc_in_row[$str_row]++;
    }
    if ( $str_col == $term_width ) {
        $str_col = 0;
        $str_row++;
    }
    $str->pos( $tmp_pos );
    #############################################
    $self->{plugin}->__restore_cursor_position();
    if ( $str_row ) {
        print "\n" x $str_row;
        $self->{plugin}->__up( $str_row );
    }
    $self->{plugin}->__clear_output( $self->{prev_str_cols} );
    $self->{plugin}->__save_cursor_position();
    if ( $opt->{no_echo} ) {
        if ( $opt->{no_echo} == 2 ) {
            print $prompt;
            return;
        }
        my $gcs_prompt = Unicode::GCString->new( $prompt );
        print $prompt . ( '*' x ( $str->length() - $gcs_prompt->length() ) );
    }
    else {
        print $str->as_string;
    }
    #############################################
    my $str_before_cursor = $str->substr( 0, $pos_str );
    my $cursor_row = 0;
    my $cursor_col = 0;
    my $gc_sum  = 0;
    for my $row ( 0 .. $#gc_in_row ) {
        if ( $gc_sum + $gc_in_row[$row] < $pos_str ) {
            $gc_sum += $gc_in_row[$row];
        }
        else {
            $cursor_row = $row;
            last;
        }
    }
    $str_before_cursor->pos( $gc_sum );
    my $gc_cursor_row = 0;
    while ( defined( my $gc = $str_before_cursor->next ) ) {
        $cursor_col += $gc->columns();
        $gc_cursor_row++;
    }
    if ( $cursor_col == $term_width ) {
        $cursor_col = 0;
        $cursor_row++;
    }
    elsif ( defined $gc_in_row[$cursor_row + 1] && $gc_cursor_row == $gc_in_row[$cursor_row] ) {
        $cursor_col = 0;
        $cursor_row++;
    }
    if ( $^O eq 'MSWin32' ) {
        $self->{prev_str_cols} = $str->columns();
        my ( $abs_col, $abs_row ) = $self->{plugin}->__get_cursor_position();
        $self->{plugin}->__set_cursor_position( $cursor_col + 1, $cursor_row + $abs_row - $str_row );
        return;
    }
    if ( $str_row > $cursor_row && $str_col == 0 ) {
        --$str_row;
        $str_col = $term_width - 1;
        if ( $self->{manually_right} && $cursor_col == $term_width - 1 ) {
            $self->{plugin}->__right( $term_width - 1 );
        }
    }
    if ( $str_row > $cursor_row ) {
        $self->{plugin}->__up( $str_row - $cursor_row );
    }
    if ( $str_col > $cursor_col ) {
        $self->{plugin}->__left( $str_col - $cursor_col );
    }
    elsif ( $str_col < $cursor_col ) {
        $self->{plugin}->__right( $cursor_col - $str_col );
    }
}




1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Term::ReadLine::Tiny - This module is now called Term::ReadLine::Simple.

=head1 VERSION

Version 0.011

=cut

=head1 SYNOPSIS

    use Term::ReadLine::Tiny;

    my $new = Term::ReadLine::Tiny->new( 'name' );
    my $line = $new->readline( 'Prompt: ', { default => 'abc' } );

=head1 DESCRIPTION

After reading the discussion in L<Please don't use ::Tiny unless it meets the ::Tiny criteria, thanks!|http://blogs.perl.org/users/dan_muey/2014/08/please-dont-use-tiny-unless-it-meets-the-tiny-criteria-thanks.html> I have decided to rename this module. The module is now called L<Term::ReadLine::Simple>. C<Term::Readline::Tiny> will be removed.

C<readline> reads a line from STDIN. As soon as C<Return> is pressed C<readline> returns the read string without the
newline character - so no C<chomp> is required.

This module is intended to cope with Unicode (multibyte character/grapheme cluster).

=head2 Keys

C<BackSpace> or C<Strg-H>: Delete the character behind the cursor.

C<Delete> or C<Strg-D>: Delete  the  character at point. Return nothing if the input puffer is empty.

C<Strg-U>: Delete the text backward from the cursor to the beginning of the line.

C<Strg-K>: Delete the text from the cursor to the end of the line.

C<Right-Arrow> or C<Strg-F>: Move forward a character.

C<Left-Arrow> or C<Strg-B>: Move back a character.

C<Home> or C<Strg-A>: Move to the start of the line.

C<End> or C<Strg-E>: Move to the end of the line.

=head1 METHODS

=head2 new

The C<new> method returns a C<Term::ReadLine::Tiny> object.

    my $new = Term::ReadLine::Tiny->new( 'name' );

The argument is the name of the application.

=head2 config

The method C<config> sets the defaults for the current C<Term::ReadLine::Tiny> object.

    $new->config( \%options );

The available options are:

=over

=item

default

Sets the default I<default> string.

Allowed values: a decoded string.

Default: not set.

=item

no_echo

Sets the default value for I<no_echo>.

Allowed values: 0, 1 or 2.

Default: 0.

=back

Options not available in the C<readline> method:

=over

=item

compat

If I<compat> is set to 1, the return value of C<readline> is not decoded else the return value of C<readline>
is decoded.

Setting the environment variable READLINE_TINY_COMPAT to a true value has the same effect as setting I<compat> to 1
unless I<compat> is defined. If I<compat> is defined, READLINE_TINY_COMPAT has no meaning.

Allowed values: 0 or 1.

Default: no set

=item

reinit_encoding

To get the right encoding C<Term::ReadLine::Tiny> uses L<Encode::Locale>. Passing an encoding to I<reinit_encoding>
changes the encoding reported by C<Encode::Locale>. See L<Encode::Locale/reinit-encoding> for more details.

Allowed values: an encoding which is recognized by the L<Encode> module.

Default: not set.

=back

=head2 readline

C<readline> reads a line from STDIN.

    $line = $new->readline( $prompt, [ \%options ] );

The fist argument is the prompt string. The optional second argument is the default string if it is not a reference. If
the second argument is a hash-reference, the hash is used to set the different options. The keys/options are

=over

=item

default

Sets a initial value of input.

=item

no_echo

If I<no_echo> is set to 1, "C<*>" are displayed instead of the characters.

If I<no_echo> is set to 2, no output is shown apart from the prompt string.

=back

See L</config> for the default and allowed values.

=head1 REQUIREMENTS

=head2 Perl version

Requires Perl version 5.08.0 or greater.

=head2 Terminal

It is required a terminal which uses a monospaced font and understands ANSI escape sequences.

=head2 Encoding layer

It is required to use appropriate I/O encoding layers. If the encoding layer for STDIN doesn't match the terminal's
character set, C<readline> will break if a non ascii character is entered.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Term::ReadLine::Tiny

=head1 AUTHOR

Matthäus Kiem <cuer2s@gmail.com>

=head1 CREDITS

Thanks to the L<Perl-Community.de|http://www.perl-community.de> and the people form
L<stackoverflow|http://stackoverflow.com> for the help.

=head1 LICENSE AND COPYRIGHT

Copyright 2014 Matthäus Kiem.

This library is free software; you can redistribute it and/or modify it under the same terms as Perl 5.10.0. For
details, see the full text of the licenses in the file LICENSE.

=cut
